lib/bundler/cli/pristine.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+require "bundler/cli/exec"
+module Bundler
+  class CLI::Pristine
+    class ConflictingGems < PristineError
+    end
+    class AmbiguousOption < PristineError
+    end
+    attr_reader :options, :gem_list
+
+    def initialize(options, gem_list)
+      @options = options
+      @gem_list = gem_list
+    end
+
+    def run
+      # Determine if -extensions and --no-extensions or  are conflict
+      if options["extensions"] && options["no-extensions"]
+        Bundler.ui.error "You can't use --extensions and --no-extensions at the same time"
+        raise AmbiguousOption
+      end
+
+      # Raise error when Gemfile is not present
+      definition = begin
+        Bundler.definition
+      rescue GemfileNotFound
+        Bundler.ui.error "You can't pristine without a Gemfile. Please consider add Gemfile then run `bundle install` before pristine"
+        raise
+      end
+      # Raise error when lockfile is not present
+      unless Bundler.default_lockfile.exist?
+        Bundler.ui.error "You can't pristine without a Gemfile.lock. Please consider run `bundle install` before pristine"
+        raise GemfileLockNotFound, "Could not locate Gemfile.lock"
+      end
+
+      # Find the path gems and exclude them from calling gem pristine
+      # If gem list is empty, we need to add everything to pristine_gems, else we just use things from the gem_list
+      if gem_list.empty?
+        pristine_gems = definition.calculate_non_path_gem_list
+        git_gems = definition.calculate_git_gems
+        path_gems = definition.calculate_path_only_gems
+        warn_path_gems(path_gems)
+        pristine(pristine_gems, git_gems, true)
+      else
+        pristine(gem_list)
+      end
+    end
+  end
+
+private
+
+  def check_conflicting(gem_list, skip_gems)
+    conflicting_gems = skip_gems & gem_list
+    unless conflicting_gems.empty?
+      Bundler.ui.error "You can't list a gem in both GEMLIST and --skip." \
+          "The offending gems are: #{conflicting_gems.join(", ")}."
+      raise ConflictingGems
+    end
+  end
+
+  def pristine(gems, git_gems = nil, lazy_spec_provided = false)
+    gem_list = lazy_spec_provided ? gems.map(&:name) : gems
+    check_conflicting(gem_list, options[:skip])
+    pristine_gems = gem_list - options[:skip]
```

Comment:
@RochesterinNYC 
To make line 53 & this line work I need to implement a "-" and a "&" on `SpecSet` since `gem_list` here could be an Array or an instance of `SpecSet` and options[:skip] would always be array. But I don't know if this would be somehow "bad practice" ? What's your opinion?